Numeric properties are not validated by the tracker and can cause ES to crash

<!--- Provide a general summary of the issue in the Title above -->
Using invalid values for the numeric properties of the tracker, Elastic Search will fail when processing the index because they can't be interpreted correctly. Eventually the constant exceptions can cause an OOM.
The store was hit by a burp suite scan and ended with this problem.
It is related to PR #975. I fixed it by removing the data for the day and adding an event processor that converts all the properties marked as numbers in elasticsuite_indices.xml.

### Preconditions
<!--- Please Provide detailed informations about the environment you use -->
Tracker is enabled

<!-- Magento Version : Are you using Magento CE or EE ? Which version of Magento are you using exactly ? -->
Magento Version : 2.3.3 CE

ElasticSearch Version: 5.6.12

<!-- ElasticSuite Version : Which exact version of ElasticSuite are you using ? -->
ElasticSuite Version : 2.8.3

<!-- Magento Environment : are you in Developer or Production mode ? -->
Environment : Production

<!-- Third party modules : are you using any third party modules ? If yes, please attach the list -->
Third party modules : not related

### Steps to reproduce
<!--- Provide a set of unambiguous steps to reproduce this bug. You can also include pieces of code if you think it's relevant  -->
1. visit on the site this url "/elasticsuite/tracker/hit/image/h.png?page[store_id]=2&page[type][identifier]=catalog_category_view&page[type][label]=Catalog%20Category&page[category][id]=507&page[category][label]=My%20Category&page[category][path]=1%2F175%2F507&page[product_list][page_count]=65&page[product_list][product_count]=645&page[product_list][current_page]=1&page[product_list][sort_order]=topseller&page[product_list][sort_direction]=asc&page[product_list][display_mode]=list&session[uid]=5550-1ba5-186a-8c8bed9dd257&session[vid]=7fd0-57d1-3e69-fd7525c4ee45&page[site]=www.example.com&page[url]=%2Fmy-category&page[title]=My%20Category&page[referrer][domain]=www.example.com&page[referrer][page]=%2Fmy-category&page[resolution][x]=128055835307'%20or%20'3426'%3d'3426&page[resolution][y]=984"  (labels and uid were changed). note the page.resolution.x value
2. in the database, the invalid string will be saved 

### Expected result
<!--- Tell us what should happen -->
1. Everything works normally

### Actual result
<!--- Tell us what happens instead -->
1. this appears constantly on the ES logs.
```
[DEBUG][o.e.a.b.TransportShardBulkAction] [3RmGevj] [magento2-prod_example_tracking_log_event_20200124][0] failed to execute bulk item (index) BulkShardRequest [[magento2-prod_example_tracking_log_event_20200124][0]] containing [1792] requests
org.elasticsearch.index.mapper.MapperParsingException: failed to parse [page.resolution.x]
[... trace]
Caused by: java.lang.NumberFormatException: For input string: "128055835307' or '3426'='3426"
```

<!--- (This may be platform independent comment) -->

--- a/src/module-elasticsuite-tracker/Model/Event/DotObject.php
+++ b/src/module-elasticsuite-tracker/Model/Event/DotObject.php
@@ -1,0 +1,200 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\Elasticsuite
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event;
+
+use Magento\Framework\DataObject;
+
+/**
+ * Variation of the DataObject to support dot notation for
+ * - testing existence
+ * - getting data
+ * - setting data
+ *
+ * @SuppressWarnings(PHPMD.CountInLoopExpression)
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class DotObject extends DataObject
+{
+    /**
+     * Overwrite data in the object.
+     *
+     * The $key parameter can be string or array.
+     * If $key is string, the attribute value will be overwritten by $value
+     * If $key is an array, it will overwrite all the data in the object.
+     *
+     * Added support for "dot" notation, using inspiration from Laravel array_set helper.
+     *
+     * @SuppressWarnings(PHPMD.CountInLoopExpression)
+     *
+     * @param string|array $key   Key where to overwrite data.
+     * @param mixed        $value Value to set for the given key.
+     *
+     * @return $this
+     */
+    public function setData($key, $value = null)
+    {
+        if ($key === (array) $key) {
+            $this->_data = $key;
+
+            return $this;
+        }
+
+        $array = &$this->_data;
+
+        $keys = explode('.', $key);
+        // @codingStandardsIgnoreStart
+        while (count($keys) > 1) {
+            $key = array_shift($keys);
+
+            // If the key doesn't exist at this depth, we will just create an empty array
+            // to hold the next value, allowing us to create the arrays to hold final
+            // values at the correct depth. Then we'll keep digging into the array.
+            if (!isset($array[$key]) || !is_array($array[$key])) {
+                $array[$key] = [];
+            }
+
+            $array = &$array[$key];
+        }
+        // @codingStandardsIgnoreEnd
+
+        $array[array_shift($keys)] = $value;
+
+        return $this;
+    }
+
+    /**
+     * Object data getter
+     *
+     * If $key is not defined will return all the data as an array.
+     * Otherwise it will return value of the element specified by $key.
+     * It is possible to use keys like a/b/c for access nested array data
+     *
+     * If $index is specified it will assume that attribute data is an array
+     * and retrieve corresponding member. If data is the string - it will be explode
+     * by new line character and converted to array.
+     *
+     * Added support for "dot" notation.
+     *
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     *
+     * @param string     $key   Data key.
+     * @param string|int $index Data index.
+     * @return mixed
+     */
+    public function getData($key = '', $index = null)
+    {
+        if ('' === $key) {
+            return $this->_data;
+        }
+
+        /* process a/b/c key as ['a']['b']['c'] */
+        if (strpos($key, '/') !== false) {
+            $data = $this->getDataByPath($key);
+        } elseif (strpos($key, '.') !== false) {
+            /* process a.b.c key as ['a']['b']['c'] */
+            $data = $this->getDataByPath($key, '.');
+        } else {
+            $data = $this->_getData($key);
+        }
+
+        if ($index !== null) {
+            if ($data === (array) $data) {
+                $data = isset($data[$index]) ? $data[$index] : null;
+            } elseif (is_string($data)) {
+                $data = explode(PHP_EOL, $data);
+                $data = isset($data[$index]) ? $data[$index] : null;
+            } elseif ($data instanceof \Magento\Framework\DataObject) {
+                $data = $data->getData($index);
+            } else {
+                $data = null;
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Get object data by path
+     *
+     * Method consider the path as chain of keys: a/b/c => ['a']['b']['c']
+     * or a path made using "do" notation: a.b.c => ['a']['b']['c']
+     *
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     *
+     * @param string $path      Path to data.
+     * @param string $delimiter Path delimiter.
+     *
+     * @return mixed
+     */
+    public function getDataByPath($path, $delimiter = '/')
+    {
+        $keys = explode($delimiter, $path);
+
+        $data = $this->_data;
+        foreach ($keys as $key) {
+            if ((array) $data === $data && isset($data[$key])) {
+                $data = $data[$key];
+            } elseif ($data instanceof \Magento\Framework\DataObject) {
+                $data = $data->getDataByKey($key);
+            } else {
+                return null;
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * If $key is empty, checks whether there's any data in the object
+     *
+     * Otherwise checks if the specified attribute is set.
+     *
+     * Added support for "dot" notation.
+     *
+     * @SuppressWarnings(PHPCS.Squiz.PHP.DisallowSizeFunctionsInLoops)
+     * @SuppressWarnings(PHPMD.CountInLoopExpression)
+     *
+     * @param string $key Data key.
+     *
+     * @return bool
+     */
+    public function hasData($key = '')
+    {
+        if (empty($key) || !is_string($key)) {
+            return !empty($this->_data);
+        }
+
+        $array = &$this->_data;
+
+        $keys = explode('.', $key);
+        // @codingStandardsIgnoreStart
+        while (count($keys) > 1) {
+            $key = array_shift($keys);
+
+            // If the key doesn't exist at this depth, then the whole structure does not exists.
+            if (!isset($array[$key]) || !is_array($array[$key])) {
+                return false;
+            }
+
+            $array = &$array[$key];
+        }
+        // @codingStandardsIgnoreEnd
+
+        return array_key_exists(array_shift($keys), $array);
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/Enforcer.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/Enforcer.php
@@ -1,0 +1,59 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\Elasticsuite
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping;
+
+use Smile\ElasticsuiteCore\Api\Index\IndexInterface;
+
+/**
+ * Mapping enforcer.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class Enforcer
+{
+    /**
+     * @var TypeEnforcerCollector
+     */
+    private $typeEnforcerCollector;
+
+    /**
+     * Enforcer constructor.
+     *
+     * @param TypeEnforcerCollector $typeEnforcerCollector Type enforcers collector.
+     */
+    public function __construct(TypeEnforcerCollector $typeEnforcerCollector)
+    {
+        $this->typeEnforcerCollector = $typeEnforcerCollector;
+    }
+
+    /**
+     * Enforces an index mapping on given data.
+     *
+     * @param IndexInterface $index Index whose mapping to enforce on data.
+     * @param array          $data  Data to enforce mapping on.
+     *
+     * @return array
+     */
+    public function enforce($index, $data)
+    {
+        $enforcers = $this->typeEnforcerCollector->getTypeEnforcers($index);
+        foreach ($enforcers as $enforcer) {
+            $data = $enforcer->enforce($data);
+        }
+
+        return $data;
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/AbstractTypeEnforcer.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/AbstractTypeEnforcer.php
@@ -1,0 +1,168 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\ElasticsuiteTracker
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer;
+
+use Smile\ElasticsuiteCore\Api\Index\Mapping\FieldInterface;
+use Smile\ElasticsuiteCore\Api\Index\MappingInterface;
+use Smile\ElasticsuiteTracker\Model\Event\DotObject;
+use Smile\ElasticsuiteTracker\Model\Event\DotObjectFactory;
+use Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcerInterface;
+
+/**
+ * Abstract field type enforcer.
+ *
+ * @category Smile.
+ * @package  Smile\ElasticsuiteTracker
+ */
+abstract class AbstractTypeEnforcer implements TypeEnforcerInterface
+{
+    /**
+     * @var DotObjectFactory
+     */
+    protected $dotObjectFactory;
+
+    /**
+     * @var MappingInterface
+     */
+    protected $mapping;
+
+    /**
+     * @var FieldInterface[]
+     */
+    protected $fields;
+
+    /**
+     * Integer constructor.
+     *
+     * @param DotObjectFactory $dataObjectFactory DotObject factory.
+     * @param MappingInterface $mapping           Index mapping.
+     */
+    public function __construct(
+        DotObjectFactory $dataObjectFactory,
+        MappingInterface $mapping
+    ) {
+        $this->dotObjectFactory = $dataObjectFactory;
+        $this->mapping          = $mapping;
+        $this->fields           = [];
+        $this->collectFields();
+    }
+
+    /**
+     * Enforces value type for collected fields.
+     *
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     *
+     * @param array $event Event data.
+     *
+     * @return array
+     */
+    public function enforce($event)
+    {
+        if (!empty($this->fields)) {
+            $eventData = $this->dotObjectFactory->create(['data' => $event]);
+
+            foreach ($this->fields as $field) {
+                if ($field->isNested()) {
+                    $this->handleNestedField($eventData, $field);
+                } else {
+                    $this->handleRegularField($eventData, $field);
+                }
+            }
+
+            $event = $eventData->getData();
+        }
+
+        return $event;
+    }
+
+    /**
+     * Collect the fields to enforce from the mapping.
+     *
+     * @return void
+     */
+    abstract protected function collectFields();
+
+    /**
+     * Enforce mapping for a given field value.
+     *
+     * @param mixed $fieldValue Field value.
+     *
+     * @return void
+     */
+    abstract protected function enforceField(&$fieldValue);
+
+    /**
+     * Handles a nested field.
+     *
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     *
+     * @param DotObject      $eventData Event data.
+     * @param FieldInterface $field     Mapping nested field.
+     *
+     * @return void
+     */
+    protected function handleNestedField($eventData, $field)
+    {
+        $fieldPath = $field->getNestedPath();
+        if ($eventData->hasData($fieldPath)) {
+            $fieldName = $field->getNestedFieldName();
+            $items = $eventData->getData($fieldPath);
+            if (is_array($items)) {
+                foreach ($items as &$item) {
+                    if (array_key_exists($fieldName, $item)) {
+                        $fieldValue = $item[$fieldName];
+                        if (is_array($fieldValue)) {
+                            array_walk(
+                                $fieldValue,
+                                array($this, 'enforceField')
+                            );
+                        } else {
+                            $this->enforceField($fieldValue);
+                        }
+                        $item[$fieldName] = $fieldValue;
+                    }
+                }
+                $eventData->setData($fieldPath, $items);
+            }
+        }
+    }
+
+    /**
+     * Handles a regular field.
+     *
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     *
+     * @param DotObject      $eventData Event data.
+     * @param FieldInterface $field     Mapping regular field.
+     *
+     * @return void
+     */
+    protected function handleRegularField($eventData, $field)
+    {
+        $fieldName = $field->getName();
+        if ($eventData->hasData($fieldName)) {
+            $fieldValue = $eventData->getData($fieldName);
+            if (is_array($fieldValue)) {
+                array_walk(
+                    $fieldValue,
+                    array($this, 'enforceField')
+                );
+            } else {
+                $this->enforceField($fieldValue);
+            }
+            $eventData->setData($fieldName, $fieldValue);
+        }
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Boolean.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Boolean.php
@@ -1,0 +1,46 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\Elasticsuite
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer;
+
+use Smile\ElasticsuiteCore\Api\Index\Mapping\FieldInterface;
+
+/**
+ * Boolean field type enforcer.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class Boolean extends AbstractTypeEnforcer
+{
+    /**
+     * {@inheritDoc}
+     */
+    protected function collectFields()
+    {
+        foreach ($this->mapping->getFields() as $field) {
+            if ($field->getType() === FieldInterface::FIELD_TYPE_BOOLEAN) {
+                $this->fields[] = $field;
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function enforceField(&$fieldValue)
+    {
+        $fieldValue = (bool) $fieldValue;
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Double.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Double.php
@@ -1,0 +1,46 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\Elasticsuite
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer;
+
+use Smile\ElasticsuiteCore\Api\Index\Mapping\FieldInterface;
+
+/**
+ * Double field type enforcer.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class Double extends AbstractTypeEnforcer
+{
+    /**
+     * {@inheritDoc}
+     */
+    protected function collectFields()
+    {
+        foreach ($this->mapping->getFields() as $field) {
+            if ($field->getType() === FieldInterface::FIELD_TYPE_DOUBLE) {
+                $this->fields[] = $field;
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function enforceField(&$fieldValue)
+    {
+        $fieldValue = (float) $fieldValue;
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Integer.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcer/Integer.php
@@ -1,0 +1,46 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\ElasticsuiteTracker
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer;
+
+use Smile\ElasticsuiteCore\Api\Index\Mapping\FieldInterface;
+
+/**
+ * Integer field type enforcer.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class Integer extends AbstractTypeEnforcer
+{
+    /**
+     * {@inheritDoc}
+     */
+    protected function collectFields()
+    {
+        foreach ($this->mapping->getFields() as $field) {
+            if ($field->getType() === FieldInterface::FIELD_TYPE_INTEGER) {
+                $this->fields[] = $field;
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function enforceField(&$fieldValue)
+    {
+        $fieldValue = (int) $fieldValue;
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcerCollector.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcerCollector.php
@@ -1,0 +1,67 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\ElasticsuiteTracker
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping;
+
+use Smile\ElasticsuiteCore\Api\Index\IndexInterface;
+
+/**
+ * Type enforcers collector.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+class TypeEnforcerCollector
+{
+    /**
+     * @var array
+     */
+    private $factories;
+
+    /**
+     * @var array
+     */
+    private $cache;
+
+    /**
+     * Enforcer constructor.
+     *
+     * @param array $factories Type enforcer factories.
+     */
+    public function __construct($factories = [])
+    {
+        $this->factories = $factories;
+        $this->cache = [];
+    }
+
+    /**
+     * Collects type enforcers for a given index mapping.
+     *
+     * @param IndexInterface $index Index.
+     *
+     * @return TypeEnforcerInterface[]
+     */
+    public function getTypeEnforcers(IndexInterface $index)
+    {
+        if (!array_key_exists($index->getIdentifier(), $this->cache)) {
+            $enforcers = [];
+            foreach ($this->factories as $enforcerFactory) {
+                $enforcers[] = $enforcerFactory->create(['mapping' => $index->getMapping()]);
+            }
+            $this->cache[$index->getIdentifier()] = $enforcers;
+        }
+
+        return $this->cache[$index->getIdentifier()];
+    }
+}

--- a/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcerInterface.php
+++ b/src/module-elasticsuite-tracker/Model/Event/Mapping/TypeEnforcerInterface.php
@@ -1,0 +1,33 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade this module to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\Elasticsuite
+ * @author    Richard BAYET <richard.bayet@smile.fr>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+
+namespace Smile\ElasticsuiteTracker\Model\Event\Mapping;
+
+/**
+ * Enforcers interface.
+ *
+ * @category Smile
+ * @package  Smile\ElasticsuiteTracker
+ */
+interface TypeEnforcerInterface
+{
+    /**
+     * Enforce a given mapping field type on event data.
+     *
+     * @param array $data Event data.
+     *
+     * @return array Event data with the mapping enforced.
+     */
+    public function enforce($data);
+}

--- a/src/module-elasticsuite-tracker/Model/EventIndex.php
+++ b/src/module-elasticsuite-tracker/Model/EventIndex.php
@@ -15,6 +15,7 @@
 namespace Smile\ElasticsuiteTracker\Model;
 
 use Smile\ElasticsuiteTracker\Api\EventIndexInterface;
+use Smile\ElasticsuiteTracker\Model\Event\Mapping\Enforcer as MappingEnforcer;
 
 /**
  * Event index implementation.
@@ -36,17 +37,25 @@ class EventIndex implements EventIndexInterface
     private $indexOperation;
 
     /**
+     * @var MappingEnforcer
+     */
+    private $mappingEnforcer;
+
+    /**
      * Constructor.
      *
-     * @param IndexResolver                                             $indexResolver  Resource model.
-     * @param \Smile\ElasticsuiteCore\Api\Index\IndexOperationInterface $indexOperation Index operation.
+     * @param IndexResolver                                             $indexResolver   Resource model.
+     * @param \Smile\ElasticsuiteCore\Api\Index\IndexOperationInterface $indexOperation  Index operation.
+     * @param MappingEnforcer                                           $mappingEnforcer Mapping enforcer.
      */
     public function __construct(
         IndexResolver $indexResolver,
-        \Smile\ElasticsuiteCore\Api\Index\IndexOperationInterface $indexOperation
+        \Smile\ElasticsuiteCore\Api\Index\IndexOperationInterface $indexOperation,
+        MappingEnforcer $mappingEnforcer
     ) {
         $this->indexResolver  = $indexResolver;
         $this->indexOperation = $indexOperation;
+        $this->mappingEnforcer = $mappingEnforcer;
     }
 
     /**
@@ -69,6 +78,7 @@ class EventIndex implements EventIndexInterface
             if (isset($event['page']['store_id'])) {
                 $index = $this->indexResolver->getIndex(self::INDEX_IDENTIFIER, $event['page']['store_id'], $event['date']);
                 if ($index !== null) {
+                    $event = $this->mappingEnforcer->enforce($index, $event);
                     $indices[$index->getName()] = $index;
                     $bulk->addDocument($index, $index->getDefaultSearchType(), $event['event_id'], $event);
                 }

--- a/src/module-elasticsuite-tracker/etc/di.xml
+++ b/src/module-elasticsuite-tracker/etc/di.xml
@@ -35,6 +35,15 @@
     <type name="Magento\Quote\Model\Quote">
       <plugin name="trackAddedProduct" type="Smile\ElasticsuiteTracker\Plugin\QuotePlugin" sortOrder="1" />
     </type>
-    
+
+    <type name="Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcerCollector">
+        <arguments>
+            <argument name="factories" xsi:type="array">
+                <item name="integer" xsi:type="object">Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer\IntegerFactory</item>
+                <item name="double" xsi:type="object">Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer\DoubleFactory</item>
+                <item name="boolean" xsi:type="object">Smile\ElasticsuiteTracker\Model\Event\Mapping\TypeEnforcer\BooleanFactory</item>
+            </argument>
+        </arguments>
+    </type>
 
 </config>

